src/node_contextify.cc

```diff
@@ -1041,38 +1072,43 @@ void ContextifyContext::CompileFunction(
     }
   }
 
-  MaybeLocal<Function> maybe_fun = ScriptCompiler::CompileFunctionInContext(
+  MaybeLocal<Function> maybe_fn = ScriptCompiler::CompileFunctionInContext(
       parsing_context, &source, params.size(), params.data(),
       context_extensions.size(), context_extensions.data(), options);
 
-  Local<Function> fun;
-  if (maybe_fun.IsEmpty() || !maybe_fun.ToLocal(&fun)) {
+  if (maybe_fn.IsEmpty()) {
     DecorateErrorStack(env, try_catch);
     try_catch.ReThrow();
     return;
   }
+  Local<Function> fn = maybe_fn.ToLocalChecked();
+  env->id_to_function_map[id].Reset(isolate, fn);
```

Comment:
It would be better to use `.emplace()` instead of relying on the implicit insertion of `[]` (`CompileFnEntry` could also be wrapped into a `BaseObject` like `ContextifyScript` so that it can be properly captured in the heap snapshot for debugging.

Reply:
I must admit this is where I have a C++/v8 knowledge gap - if I try `env->id_to_function_map.emplace` then, as far as I recall, the copy constructor error for Persistent kicks in and we can't use Persistent and instead need to use Persistent with copyable traits, which I know is just a change of the Persistent constructor but it's a truly ugly API and I wasn't sure how to do that in the above form so this seemed easier to avoid the copy constructor entirely.

Reply:
(for what its worth this pattern comes from module_wrap.cc)

Reply:
I think the idea would be to do something like this:

```suggestion
  env->id_to_function_map.emplace(std::piecewise_construct,
                                  std::make_tuple(id),
                                  std::make_tuple(isolate, fn));
```